### PR TITLE
Refactor tests.testDCClient as dynamicconfig.MemoryClient

### PR DESCRIPTION
**Pull request by prathyushpv**
## What changed?
Refactor testDCClient from tests package as dynamicconfig.MemoryClient.

## Why?
This DC Client is useful for testing in other packages.

## How did you test it?
Existing unit tests

## Potential risks
None

## Is hotfix candidate?
No

### CODE REVIEW

**Review by dnr — COMMENTED** [review]
What do you think about this instead:
- Move this to `common/dynamicconfig` and call it `MemoryClient`
- Get rid of the "fallback" feature that's not used (just return nil instead)
- Make a `SetValue[ByKey]` on `MemoryClient` not use `testing.T` (so that `MemoryClient` doesn't depend on `testing`), but then:
- Make `OverrideValue[ByKey]` on either `temporalImpl` or `FunctionalTestBase` that does hook into testing/suite to do cleanup

**Comment by prathyushpv** [comment]
> What do you think about this instead:
> 
> * Move this to `common/dynamicconfig` and call it `MemoryClient`
> * Get rid of the "fallback" feature that's not used (just return nil instead)
> * Make a `SetValue[ByKey]` on `MemoryClient` not use `testing.T` (so that `MemoryClient` doesn't depend on `testing`), but then:
> * Make `OverrideValue[ByKey]` on either `temporalImpl` or `FunctionalTestBase` that does hook into testing/suite to do cleanup

Yeah that makes more sense.
I have refactored the code to do this. I moved cleanup logic to temporalImpl. But I kept fallback. This could be useful if a test wants to read configs from a file and then override a few keys.

File: common/dynamicconfig/memory_client.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	sync.RWMutex
+	overrides map[Key]any
+	fallback  Client
+}
+
+func (d *MemoryClient) GetRawValue(name Key) (any, bool) {
+	d.RLock()
+	defer d.RUnlock()
+	v, ok := d.overrides[name]
+	return v, ok
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:41]:
```suggestion
	return d.overrides[name]
```
should work?

File: common/dynamicconfig/memory_client.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+// THE SOFTWARE.
+
+package dynamicconfig
+
+import (
+	"sync"
+)
+
+type MemoryClient struct {
+	sync.RWMutex
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:32]: I like to do
```suggestion
	lock sync.RWMutex
```
so that clients don't have direct access to the lock

File: common/dynamicconfig/memory_client.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+func (d *MemoryClient) GetValue(name Key) []ConstrainedValue {
+	if val, ok := d.GetRawValue(name); ok {
+		return []ConstrainedValue{{Value: val}}
+	}
+	return d.fallback.GetValue(name)
+}
+
+func (d *MemoryClient) OverrideValue(setting GenericSetting, value any) {
+	d.overrides[setting.Key()] = value
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:52]: missing lock?  maybe call OverrideValueByKey instead?

File: tests/advanced_visibility.go [hunk 4]
```diff
@@ -2022,9 +2022,8 @@ func (s *AdvancedVisibilitySuite) Test_BuildIdIndexedOnCompletion_UnversionedWor
 
 func (s *AdvancedVisibilitySuite) Test_BuildIdIndexedOnCompletion_VersionedWorker() {
 	// Use only one partition to avoid having to wait for user data propagation later
-	dc := s.testCluster.host.dcClient
-	dc.OverrideValue(s.T(), dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
-	dc.OverrideValue(s.T(), dynamicconfig.MatchingNumTaskqueueWritePartitions, 1)
+	s.testCluster.host.OverrideDCValue(s.T(), dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
```
Comment (dnr) [inline tests/advanced_visibility.go:2025]: I want to simplify this more so you don't need the call to T() and the long path, but that can be in another PR

File: common/dynamicconfig/memory_client.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func (d *MemoryClient) GetRawValue(name Key) (any, bool) {
+	d.RLock()
+	defer d.RUnlock()
+	v, ok := d.overrides[name]
+	return v, ok
+}
+
+func (d *MemoryClient) GetValue(name Key) []ConstrainedValue {
+	if val, ok := d.GetRawValue(name); ok {
+		return []ConstrainedValue{{Value: val}}
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:46]: I know it's not used right now, but we could do this to make it more flexible:
if `val` already has type `[]ConstrainedValue` then return it directly, otherwise wrap it like this.

File: tests/onebox.go [hunk 6]
```diff
@@ -221,11 +247,17 @@ func newTemporal(t *testing.T, params *TemporalParams) *temporalImpl {
...
 		tlsConfigProvider:                params.TLSConfigProvider,
 		captureMetricsHandler:            params.CaptureMetricsHandler,
-		dcClient:                         testDCClient,
+		dcClient:                         dynamicconfig.NewMemoryDCClient(dynamicconfig.NewNoopClient()),
 		serviceFxOptions:                 params.ServiceFxOptions,
 		taskCategoryRegistry:             params.TaskCategoryRegistry,
 	}
-	impl.overrideHistoryDynamicConfig(t, testDCClient)
+	for k, v := range staticOverrides {
+		impl.overrideDynamicConfigByKey(t, k, v)
```
Comment (dnr) [inline tests/onebox.go:255]: I'm not sure if this is better or worse, but just to consider:
The `fallback` for the memory client can be `dynamicconfig.StaticClient(staticOverrides)`, then you only need to set the ones from params and history ones.

File: common/dynamicconfig/memory_client.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (d *MemoryClient) RemoveOverrideByKey(name Key) {
+	d.Lock()
+	defer d.Unlock()
+	delete(d.overrides, name)
+}
+
+// NewMemoryDCClient - returns a memory based dynamic config client
+func NewMemoryDCClient(fallback Client) *MemoryClient {
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:72]:
```suggestion
func NewMemoryClient(fallback Client) *MemoryClient {
```
"DC" is in the package name, don't need to repeat it

File: tests/onebox.go [hunk 8]
```diff
@@ -998,3 +1030,20 @@ func sdkClientFactoryProvider(
 		dynamicconfig.WorkerStickyCacheSize.Get(dc),
 	)
 }
+
+func (c *temporalImpl) overrideDynamicConfigByKey(t *testing.T, name dynamicconfig.Key, value any) {
+	constrainedValues := c.dcClient.GetValue(name)
+	existed := len(constrainedValues) == 0
```
Comment (dnr) [inline tests/onebox.go:1036]: shouldn't this be != 0?

File: tests/onebox.go [hunk 9]
```diff
@@ -998,3 +1030,20 @@ func sdkClientFactoryProvider(
...
+	c.dcClient.OverrideValueByKey(name, value)
+	t.Cleanup(func() {
+		if existed {
+			var values []any
+			for _, v := range constrainedValues {
+				values = append(values, v)
+			}
+			c.dcClient.OverrideValueByKey(name, values)
```
Comment (dnr) [inline tests/onebox.go:1044]: doesn't this not work without the type check in GetValue? it'll re-wrap the list...
but also, why copy it to a `[]any` at all?

Comment (prathyushpv) [reply tests/onebox.go:1044]: Yeah, that was a mistake.

File: tests/onebox.go (hunk 8 above)
Comment (prathyushpv) [reply tests/onebox.go:1036]: Yes!

File: common/dynamicconfig/memory_client.go (hunk 1 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go:41]: TIL it is not possible to return d.overrides[name]. This will only return the value and not (value, ok).

File: common/dynamicconfig/memory_client.go (hunk 5 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go:46]: Done

File: common/dynamicconfig/memory_client.go (hunk 3 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go:52]: Done

File: common/dynamicconfig/memory_client.go (hunk 7 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go:72]: Done

File: tests/advanced_visibility.go (hunk 4 above)
Comment (prathyushpv) [reply tests/advanced_visibility.go:2025]: Do you propose adding OverrideDCValue() to FunctionalTestBase?

File: tests/onebox.go (hunk 8 above)
Comment (prathyushpv) [reply tests/onebox.go:1036]: Done

File: tests/onebox.go (hunk 9 above)
Comment (prathyushpv) [reply tests/onebox.go:1044]: Done

File: tests/advanced_visibility.go (hunk 4 above)
Comment (prathyushpv) [reply tests/advanced_visibility.go]: I have made that change as well. Let me know if that is what you were thinking about.

File: common/dynamicconfig/memory_client.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	overrides map[Key]any
+	fallback  Client
+}
+
+func (d *MemoryClient) GetRawValue(name Key) (any, bool) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	v, ok := d.overrides[name]
+	if value, ok := v.([]ConstrainedValue); ok {
+		return value[0].Value, true
```
Comment (prathyushpv) [inline common/dynamicconfig/memory_client.go:42]: Now that we allow either []ConstrainedValue or a Value in the map, I am confused about how GetRawValue() should behave. Tests expect it to return a value. But we might have []ConstrainedValue. Is it okay to return the first value in the list?

File: common/dynamicconfig/memory_client.go (hunk 5 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go]: Now that we allow either []ConstrainedValue or a Value in the map, I am confused about how GetRawValue() should behave. Tests expect it to return a value. But we might have []ConstrainedValue. Is it okay to return the first value in the list?

File: common/dynamicconfig/memory_client.go [hunk 11]
```diff
@@ -29,32 +29,40 @@ import (
...
 }
 
 func (d *MemoryClient) GetRawValue(name Key) (any, bool) {
-	d.RLock()
-	defer d.RUnlock()
+	d.lock.RLock()
+	defer d.lock.RUnlock()
 	v, ok := d.overrides[name]
+	if value, ok := v.([]ConstrainedValue); ok {
+		return value[0].Value, true
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:42]: this seems wrong, you can't just pick the first one.. also this whole function is unused now?

oh, I see where this is used. uugh, that's messy. I think you can just use GetValue there and get rid of the "must override" check. pick the first value from the ConstrainedValues with len(Constraints)==0, and panic if there is none.

(the idea was that those tests always should be explicit about partition count since the behavior can be pretty different with 1 vs many partitions, and I wanted most tests to use 1 partition. but I think it doesn't really matter that much.)

File: common/dynamicconfig/memory_client.go [hunk 12]
```diff
@@ -29,32 +29,40 @@ import (
...
-		return []ConstrainedValue{{Value: val}}
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	if v, ok := d.overrides[name]; ok {
+		if value, ok := v.([]ConstrainedValue); ok {
+			return value
+		}
+		return []ConstrainedValue{{Value: v}}
 	}
 	return d.fallback.GetValue(name)
```
Comment (dnr) [inline common/dynamicconfig/memory_client.go:56]: Should probably not hold the lock while calling the fallback? I guess it's unlikely to matter in practice.

But I'm also still questioning the need for the fallback. Specifically: the next step is to make MemoryClient implement NotifyingClient so subscriptions work instantaneously. Doing that with fallback is a lot more complicated since you have to check if that one can notify, subscribe to it, pass through callbacks, etc. And having a fallback but having subscriptions only work on MemoryClient but not the fallback is confusing semantics. So I might still lean towards dropping it, if there's not a very clear use case.

File: tests/functional_test_base.go [hunk 13]
```diff
@@ -437,3 +437,7 @@ func (s *FunctionalTestBase) registerArchivalNamespace(archivalNamespace string)
 	)
 	return err
 }
+
+func (s *FunctionalTestBase) overrideDynamicConfig(t *testing.T, setting dynamicconfig.GenericSetting, value any) {
```
Comment (dnr) [inline tests/functional_test_base.go:441]: great, this is a lot cleaner. I think it can be improved even more to have this call s.T() instead of having the caller do it.

File: tests/nexus_workflow_test.go [hunk 14]
```diff
@@ -45,11 +45,11 @@ import (
 	"go.temporal.io/api/workflowservice/v1"
 	"go.temporal.io/sdk/client"
 	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/server/common/dynamicconfig"
```
Comment (dnr) [inline tests/nexus_workflow_test.go:48]: move to server block

File: tests/nexus_workflow_test.go [hunk 15]
```diff
@@ -772,7 +772,7 @@ func (s *ClientFunctionalSuite) TestNexusOperationAsyncCompletionErrors() {
 	s.NoError(err)
 
 	s.T().Run("ConfigDisabled", func(t *testing.T) {
```
Comment (dnr) [inline tests/nexus_workflow_test.go:774]: use `s.Run()` instead of `s.T().Run()`, then `s.T()` will return the subtest `t` in the test, and you can safely call `s.T()` from `overrideDynamicConfig`

File: common/dynamicconfig/memory_client.go (hunk 1 above)
Comment (dnr) [reply common/dynamicconfig/memory_client.go]: doh, you're right

File: tests/advanced_visibility.go (hunk 4 above)
Comment (dnr) [reply tests/advanced_visibility.go]: yup, exactly what I was thinking, except you can move the s.T() also

File: common/dynamicconfig/memory_client.go (hunk 5 above)
Comment (dnr) [reply common/dynamicconfig/memory_client.go]: commented on the new code.. let's get rid of GetRawValue

File: tests/nexus_workflow_test.go (hunk 15 above)
Comment (prathyushpv) [reply tests/nexus_workflow_test.go:774]: Oh that was the issue! Thanks! I tried removing t argument from OverrideDCValue() and reverted that because of this test failing.

File: common/dynamicconfig/memory_client.go (hunk 5 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go]: Done

File: common/dynamicconfig/memory_client.go (hunk 12 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go]: Right. Now I think it is better to remove fallback.

File: tests/functional_test_base.go (hunk 13 above)
Comment (prathyushpv) [reply tests/functional_test_base.go]: Done

File: tests/nexus_workflow_test.go (hunk 14 above)
Comment (prathyushpv) [reply tests/nexus_workflow_test.go]: Done

File: tests/advanced_visibility.go (hunk 4 above)
Comment (prathyushpv) [reply tests/advanced_visibility.go]: Done

File: common/dynamicconfig/memory_client.go (hunk 11 above)
Comment (prathyushpv) [reply common/dynamicconfig/memory_client.go]: Done

File: tests/onebox.go [hunk 16]
```diff
@@ -998,3 +1030,15 @@ func sdkClientFactoryProvider(
 		dynamicconfig.WorkerStickyCacheSize.Get(dc),
 	)
 }
+
+func (c *temporalImpl) overrideDynamicConfigByKey(t *testing.T, name dynamicconfig.Key, value any) {
+	existingValues := c.dcClient.GetValue(name)
+	c.dcClient.OverrideValueByKey(name, value)
+	t.Cleanup(func() {
+		if len(existingValues) > 0 {
+			c.dcClient.OverrideValueByKey(name, existingValues)
```
Comment (dnr) [inline tests/onebox.go:1039]: just thinking.. if there was a static override with a plain value, this takes the return value from GetValue, and passes it back to OverrideValueByKey, it'll get wrapped in a []ConstrainedValue, so not identical to how it was before. that's totally fine though, it'll work as expected (as long as Cleanup is called like a stack, which it is).

File: tests/namespace_delete.go [hunk 17]
```diff
@@ -112,11 +112,10 @@ func (s *namespaceTestSuite) SetupTest() {
...
 
 func (s *namespaceTestSuite) Test_NamespaceDelete_InvalidUTF8() {
-	dc := s.cluster.host.dcClient
 	// don't fail for this test, we're testing this behavior specifically
-	dc.OverrideValue(s.T(), dynamicconfig.ValidateUTF8FailRPCRequest, false)
-	dc.OverrideValue(s.T(), dynamicconfig.ValidateUTF8FailRPCResponse, false)
-	dc.OverrideValue(s.T(), dynamicconfig.ValidateUTF8FailPersistence, false)
+	s.cluster.host.OverrideDCValue(s.T(), dynamicconfig.ValidateUTF8FailRPCRequest, false)
+	s.cluster.host.OverrideDCValue(s.T(), dynamicconfig.ValidateUTF8FailRPCResponse, false)
+	s.cluster.host.OverrideDCValue(s.T(), dynamicconfig.ValidateUTF8FailPersistence, false)
```
Comment (dnr) [inline tests/namespace_delete.go:118]: should these use `s.overrideDynamicConfig` also?

File: tests/functional_test_base.go [hunk 18]
```diff
@@ -437,3 +437,7 @@ func (s *FunctionalTestBase) registerArchivalNamespace(archivalNamespace string)
 	)
 	return err
 }
+
+func (s *FunctionalTestBase) overrideDynamicConfig(setting dynamicconfig.GenericSetting, value any) {
```
Comment (dnr) [inline tests/functional_test_base.go:441]: I'm wondering if this should be public so it can be used from other packages, e.g. the ndc/xdc tests. Currently some xdc tests do override dynamic config, but they manage multiple clusters so they call temporalImpl.OverrideDCValue directly, which is fine.

**Review by dnr — APPROVED** [review]
Thanks, this is great!

Just a couple more comments but feel free to fix (or not) and merge

File: tests/namespace_delete.go (hunk 17 above)
Comment (prathyushpv) [reply tests/namespace_delete.go:118]: This test suite does not use FunctionalTestBase. Probably it can be refactored to use it.

File: tests/functional_test_base.go (hunk 18 above)
Comment (prathyushpv) [reply tests/functional_test_base.go:441]: No tests in other packages were embedding FunctionalTestBase so far. But you are right, it should be possible.

File: tests/onebox.go (hunk 16 above)
Comment (prathyushpv) [reply tests/onebox.go:1039]: Yeah, it should be fine. We could wrap the value in []ConstrainedValue when it is first inserted so that it stays consistent.
